File: blueline/blueline/page/coa_console/coa_console.py
import frappe
from frappe import _
from frappe.utils import strip_html

DOCTYPE = "Company COA Configuration"

CONFIG_FIELDS = [
	"company",
	"default_cash_account",
	"default_bank_account",
	"temporary_opening_account",
	"default_cost_center",
	"stock_adjustment_account",
	"stock_received_but_not_billed",
	"expenses_included_in_valuation",
	"default_valuation_method",
	"default_warehouse",
	"exchange_gain_loss_account",
	"foreign_currency_revaluation_account",
	"inter_company_clearing_account",
	"earliest_allowed_posting_date",
]

# `company` is the document name and is set-only-once, so it is never editable from the console.
EDITABLE_FIELDS = [fieldname for fieldname in CONFIG_FIELDS if fieldname != "company"]

WRITE_ROLES = {"Financial Manager", "System Manager"}


def _check_permission():
	if not frappe.has_permission(DOCTYPE, "read"):
		frappe.throw(_("Not permitted to view {0}").format(_(DOCTYPE)), frappe.PermissionError)


@frappe.whitelist()
def get_configurations():
	_check_permission()

	# get_list (not get_all / db.sql) so the user's role permissions and Company User
	# Permissions are applied by the framework; a restricted user only receives their companies.
	configurations = frappe.get_list(
		DOCTYPE,
		fields=["name", *CONFIG_FIELDS],
		order_by="company asc",
		limit_page_length=0,
	)

	for config in configurations:
		config["can_write"] = bool(frappe.has_permission(DOCTYPE, "write", doc=config.name))

	return {
		"configurations": configurations,
		"can_write_any": bool(WRITE_ROLES & set(frappe.get_roles())),
	}


@frappe.whitelist(methods=["POST"])
def save_configuration(company, values):
	# The client-side "Editable" badge is only a hint; write permission is re-checked here
	# against the specific document (role permissions + Company User Permissions).
	doc = None
	if isinstance(company, str) and frappe.db.exists(DOCTYPE, company):
		doc = frappe.get_doc(DOCTYPE, company)

	# A missing document and a forbidden one give the same answer so the endpoint cannot be
	# used to probe which companies have a configuration.
	if not doc or not frappe.has_permission(DOCTYPE, "write", doc=doc):
		frappe.throw(
			_("Not permitted to update the COA configuration for this company"),
			frappe.PermissionError,
		)

	values = frappe.parse_json(values)
	if not isinstance(values, dict):
		frappe.throw(_("Invalid values"))

	not_editable = sorted(set(values) - set(EDITABLE_FIELDS))
	if not_editable:
		frappe.throw(_("These fields cannot be edited here: {0}").format(", ".join(not_editable)))

	for fieldname, value in values.items():
		if value is not None and not isinstance(value, str):
			frappe.throw(_("Invalid value for {0}").format(fieldname))
		doc.set(fieldname, value or None)

	# doc.save() (not db.set_value) so validate() and Track Changes run, and the framework
	# checks write permission a second time; ignore_permissions is deliberately not set.
	try:
		doc.save()
	except frappe.ValidationError as e:
		frappe.db.rollback()
		frappe.clear_messages()
		return {"ok": False, "message": strip_html(str(e))}

	configuration = {fieldname: doc.get(fieldname) for fieldname in ["name", *CONFIG_FIELDS]}
	configuration["can_write"] = True

	return {
		"ok": True,
		"message": _("Saved COA configuration for {0}").format(doc.name),
		"configuration": configuration,
	}

File: blueline/blueline/page/coa_console/coa_console.js
frappe.provide("blueline");

frappe.pages["coa-console"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("COA Configuration Console"),
		single_column: true,
	});

	wrapper.coa_console = new blueline.CoaConsole(page);
};

blueline.CoaConsole = class CoaConsole {
	constructor(page) {
		this.page = page;
		this.setup_actions();
		this.setup_body();
		this.refresh();
	}

	setup_actions() {
		this.page.set_primary_action(__("Refresh"), () => this.refresh(), "refresh");
	}

	setup_body() {
		if (!document.getElementById("coac-style")) {
			$(`<style id="coac-style">
				.coa-console { padding: 15px 0; }
				.coa-console .coac-panel-col { margin-bottom: 20px; }
				.coa-console .coac-panel {
					--coac-accent: #5e64ff;
					background: var(--card-bg, var(--fg-color));
					border: 1px solid var(--border-color);
					border-top: 4px solid var(--coac-accent);
					border-radius: var(--border-radius-md, 6px);
					height: 100%;
				}
				.coa-console .coac-panel-header {
					display: flex;
					align-items: center;
					justify-content: space-between;
					gap: 10px;
					padding: 12px 15px;
					background: color-mix(in srgb, var(--coac-accent) 10%, transparent);
					border-bottom: 1px solid var(--border-color);
				}
				.coa-console .coac-company-name {
					font-size: 16px;
					font-weight: 600;
					color: var(--text-color);
					word-break: break-word;
				}
				.coa-console .coac-badge {
					flex-shrink: 0;
					font-size: 11px;
					font-weight: 600;
					text-transform: uppercase;
					padding: 3px 8px;
					border-radius: 10px;
				}
				.coa-console .coac-badge-editable {
					background: var(--bg-green, #e4f5e9);
					color: var(--text-on-green, #2f7a4d);
				}
				.coa-console .coac-badge-readonly {
					background: var(--bg-gray, #f3f3f3);
					color: var(--text-on-gray, #6c7680);
				}
				.coa-console .coac-panel-body { padding: 5px 15px 10px; }
				.coa-console .coac-section { padding: 10px 0; }
				.coa-console .coac-section + .coac-section { border-top: 1px dashed var(--border-color); }
				.coa-console .coac-section-title {
					font-size: 12px;
					font-weight: 600;
					text-transform: uppercase;
					color: var(--coac-accent);
					margin-bottom: 6px;
				}
				.coa-console .coac-field {
					display: flex;
					justify-content: space-between;
					gap: 15px;
					padding: 3px 0;
				}
				.coa-console .coac-field-label {
					flex: 0 0 42%;
					color: var(--text-muted);
					font-size: 12px;
				}
				.coa-console .coac-field-value {
					flex: 1;
					text-align: right;
					color: var(--text-color);
					word-break: break-word;
				}
				.coa-console .coac-not-set { color: var(--text-muted); font-style: italic; }
				.coa-console .coac-empty {
					text-align: center;
					padding: 50px 15px;
					color: var(--text-muted);
				}
			</style>`).appendTo("head");
		}

		this.body = $(`
			<div class="coa-console">
				<div class="row coac-panels"></div>
			</div>
		`).appendTo(this.page.body);
	}

	refresh() {
		const me = this;
		frappe.call({
			method: "blueline.blueline.page.coa_console.coa_console.get_configurations",
			freeze: true,
			callback(r) {
				if (r.message) {
					me.render(r.message);
				}
			},
		});
	}

	render(data) {
		this.render_access_indicator(data.can_write_any);

		const configurations = data.configurations || [];
		const $panels = this.body.find(".coac-panels").empty();

		if (!configurations.length) {
			$panels.html(`
				<div class="col-md-12">
					<div class="coac-empty">${__("No COA configurations available to you")}</div>
				</div>
			`);
			return;
		}

		configurations.forEach((config, index) => {
			$panels.append(this.get_panel_html(config, index));
		});
	}

	render_access_indicator(can_write_any) {
		this.page.set_indicator(
			can_write_any ? __("Edit access") : __("View only"),
			can_write_any ? "green" : "gray"
		);
	}

	get_panel_html(config, index) {
		const accents = blueline.CoaConsole.ACCENTS;
		const accent = accents[index % accents.length];
		const badge = config.can_write
			? `<span class="coac-badge coac-badge-editable">${__("Editable")}</span>`
			: `<span class="coac-badge coac-badge-readonly">${__("Read-only")}</span>`;

		const sections = blueline.CoaConsole.SECTIONS.map(
			(section) => `
				<div class="coac-section">
					<div class="coac-section-title">${__(section.title)}</div>
					${section.fields.map((field) => this.get_field_html(config, field)).join("")}
				</div>
			`
		).join("");

		return `
			<div class="col-lg-6 col-md-12 coac-panel-col">
				<div class="coac-panel" style="--coac-accent: ${accent}">
					<div class="coac-panel-header">
						<div class="coac-company-name">${frappe.utils.escape_html(config.company || config.name)}</div>
						${badge}
					</div>
					<div class="coac-panel-body">${sections}</div>
				</div>
			</div>
		`;
	}

	get_field_html(config, field) {
		const raw = config[field.fieldname];
		let value = `<span class="coac-not-set">${__("Not set")}</span>`;

		if (raw) {
			const text = field.type === "date" ? frappe.datetime.str_to_user(raw) : raw;
			value = frappe.utils.escape_html(text);
		}

		return `
			<div class="coac-field">
				<span class="coac-field-label">${__(field.label)}</span>
				<span class="coac-field-value">${value}</span>
			</div>
		`;
	}
};

blueline.CoaConsole.ACCENTS = ["#5e64ff", "#28a745", "#ffa00a", "#ff5858", "#743ee2", "#00b0af", "#7cd6fd"];

blueline.CoaConsole.SECTIONS = [
	{
		title: "Financial & Balancing",
		fields: [
			{ fieldname: "default_cash_account", label: "Default Cash Account" },
			{ fieldname: "default_bank_account", label: "Default Bank Account" },
			{ fieldname: "temporary_opening_account", label: "Temporary Opening Account" },
			{ fieldname: "default_cost_center", label: "Default Cost Center" },
		],
	},
	{
		title: "Stock & Valuation",
		fields: [
			{ fieldname: "stock_adjustment_account", label: "Stock Adjustment Account" },
			{ fieldname: "stock_received_but_not_billed", label: "Stock Received But Not Billed" },
			{ fieldname: "expenses_included_in_valuation", label: "Expenses Included In Valuation" },
			{ fieldname: "default_valuation_method", label: "Default Valuation Method" },
			{ fieldname: "default_warehouse", label: "Default Warehouse" },
		],
	},
	{
		title: "Multi-Currency",
		fields: [
			{ fieldname: "exchange_gain_loss_account", label: "Exchange Gain/Loss Account" },
			{
				fieldname: "foreign_currency_revaluation_account",
				label: "Foreign Currency Revaluation Account",
			},
			{ fieldname: "inter_company_clearing_account", label: "Inter Company Clearing Account" },
		],
	},
	{
		title: "Guardrail",
		fields: [
			{
				fieldname: "earliest_allowed_posting_date",
				label: "Earliest Allowed Posting Date",
				type: "date",
			},
		],
	},
];
